### DISCUSSION TITLE
(v2.2.3) How to use training loop with JSON data?

### DISCUSSION THREAD

**Discussion by abs-xyz** [explosion/spaCy#10883] [Help: Coding & Implementations]
This is regarding spaCy v2 (specifically 2.2.3). I need to change some hyperparameters.
I have seen the examples [here](https://v2.spacy.io/usage/training). They use nlp.update() to set certain things. But the training data is in the form
```
("I like London and Berlin.", {"entities": [(7, 13, "LOC"), (18, 24, "LOC")]})
```
I have JSON format data (converted from CoNLL format using `spacy convert`). How can I set values of hyperparameters?

I saw [this](https://github.com/explosion/spaCy/discussions/5671) discussion. This shows how to change the dropout rate and batch sizes, but I can't find how to change the learning rate, etc.

(My main question is on how to change hyperparameters. One of the ways in some discussion was to use the training loop, but I am not sure on how to run it with the data format that I have.)

Edit:
If I do something like `learn_rate = spacy.util.env_opt("learn_rate", 0.0001)`, will this learning rate be used while training with `spacy train`? (How do I confirm this? `os.environ["SPACY_LEARN_RATE"]` gives a KeyError.)
Also, doing something like `!learn_rate=0.0001 python -m spacy train ...` doesn't work (in a notebook).

**Comment by rmitsch** [comment]
The [v2 documentation on training your models](https://v2.spacy.io/api/cli#train) describes changing your hyperparameters. I'd recommend switching to v3 however, since configuring this is more [convenient there](https://spacy.io/usage/v3#features-training).
Also, can you elaborate how the format of your training data relates to your question about hyperparameters?

**Reply by abs-xyz** [reply]
If I am not mistaken, there is no flag to change the `learn_rate`.
Does using `set learn_rate=0.0001` on CMD change the learn rate? Is there some way to "confirm" this?

**Reply by abs-xyz** [reply]
Also, is the correct way `os.environ["LEARN_RATE"]="0.0001"` or `os.environ["SPACY_LEARN_RATE"]="0.0001"`? (I think the latter is, from [here](https://github.com/explosion/spaCy/blob/c94f7d0e75e9e4ce25b719edee3adb4ecd74ee50/spacy/util.py#L354))

**Reply by abs-xyz** [reply]
>Also, can you elaborate how the format of your training data relates to your question about hyperparameters?

[Here](https://github.com/explosion/spaCy/issues/3352#issuecomment-491515120) it is told to use the `nlp.update()` function.
[From the v2 docs](https://v2.spacy.io/api/language#update),
```
for raw_text, entity_offsets in train_data:
    doc = nlp.make_doc(raw_text)
    gold = GoldParse(doc, entities=entity_offsets)
    nlp.update([doc], [gold], drop=0.5, sgd=optimizer)
```
My question: How do I use the JSON data in this loop (for `train_data`)? It seems to be of the format that I mentioned in my original post.

**Reply by rmitsch** [reply]
Before we get deeper into this, are you starting from scratch or do you already have a substantial code base in v2?
